### PR TITLE
Split built-in and custom algorithm manifests to fix upgrading issue

### PR DESCRIPTION
**Pull request by liuzhe-lz**
### Description ###

This is a rewrite of PR #4367, following Yuqing's suggestion.

### Checklist ###
  - [ ] ~~test case~~ (covered by IT)
  - [x] doc

### How to test ###

1. remove ~/.config/nni
2. install an old version
3. register a custom tuner
4. upgrade to latest version
5. test Random/TPE with new args (e.g. seed)
6. test custom tuner

### CODE REVIEW

File: nni/tools/package_utils/config_manager.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    with open(path) as f:
+        config = yaml.safe_load(f)
+    algos = []
+    for algo_type in ['tuner', 'assessor', 'advisor']:
+        for algo in config.get(algo_type + 's', []):
+            algos.append(AlgoMeta.load(algo, algo_type))
+    return algos
+
+def _save_custom_config(algos):
+    config = defaultdict(list)
```
Comment (ultmaster) [inline nni/tools/package_utils/config_manager.py:71]: Still dumping all the algos, including builtin?

File: nni/tools/package_utils/common.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+from typing import Dict, NamedTuple, Optional
+
+class AlgoMeta(NamedTuple):
+    name: str
+    class_name: Optional[str]
+    accept_class_args: bool
+    class_args: Optional[dict]
+    validator_class_name: Optional[str]
+    algo_type: str
```
Comment (ultmaster) [inline nni/tools/package_utils/common.py:14]: `Literal['tuner', 'assessor', 'advisor']`

Comment (cruiseliu) [reply nni/tools/package_utils/common.py:14]: Literal was introduced in Python 3.8.

File: nni/tools/package_utils/config_manager.py (hunk 1 above)
Comment (cruiseliu) [reply nni/tools/package_utils/config_manager.py:71]: The parameter `algos` should only contain custom algos. I will update the param name.

**Review by ultmaster — APPROVED** [review]

File: nni/tools/package_utils/common.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+from typing import Dict, NamedTuple, Optional
+
+class AlgoMeta(NamedTuple):
+    name: str
+    class_name: Optional[str]
+    accept_class_args: bool
+    class_args: Optional[dict]
+    validator_class_name: Optional[str]
+    algo_type: str  # 'tuner' | 'assessor' | 'advisor'
+    is_builtin: bool
```
Comment (QuanluZhang) [inline nni/tools/package_utils/common.py:15]: should we add version for each algorithm, at least for customized algorithm?

Comment (liuzhe-lz) [reply nni/tools/package_utils/common.py:15]: Tuner does not have version.

**Review by QuanluZhang — APPROVED** [review]

**Pull request merged**